openapi/src/main/java/io/helidon/openapi/ExpandedTypeDescription.java

```diff
@@ -0,0 +1,303 @@
+
+                default:
+
+            }
+        }
+        return false;
+    }
+
+    private boolean isExtension(String name) {
+        return name.startsWith(EXTENSION_PROPERTY_PREFIX);
+    }
+
+    /**
+     * Specific type description for {@code Schema}.
+     * <p>
+     *     The {@code Schema} node allows the {@code additionalProperties} subnode to be either
+     *     {@code Boolean} or another {@code Schema}. This type description provides a customized
+     *     property description for {@code additionalProperties} that infers which variant a
+     *     specific node in the document actually uses and then processes it accordingly.
+     * </p>
+     */
+    static class SchemaTypeDescription extends ExpandedTypeDescription {
+
+        private static final PropertyDescriptor ADDL_PROPS_PROP_DESCRIPTOR = preparePropertyDescriptor();
+
+        private static final Property ADDL_PROPS_PROPERTY =
+                new MethodProperty(ADDL_PROPS_PROP_DESCRIPTOR) {
+
+                    @Override
+                    public void set(Object object, Object value) throws Exception {
+                        Schema s = Schema.class.cast(object);
+                        if (value instanceof Schema) {
+                            s.setAdditionalPropertiesSchema((Schema) value);
+                        } else {
+                            s.setAdditionalPropertiesBoolean((Boolean) value);
+                        }
+                    }
+
+                    @Override
+                    public Object get(Object object) {
+                        Schema s = Schema.class.cast(object);
+                        Boolean b = s.getAdditionalPropertiesBoolean();
+                        return b != null ? b : s.getAdditionalPropertiesSchema();
+                    }
+                };
+
+        private static PropertyDescriptor preparePropertyDescriptor() {
+            try {
+                return new PropertyDescriptor("additionalProperties",
+                        Schema.class.getMethod("getAdditionalPropertiesSchema"),
+                        Schema.class.getMethod("setAdditionalPropertiesSchema", Schema.class));
+            } catch (IntrospectionException | NoSuchMethodException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private SchemaTypeDescription(Class<? extends Object> clazz, Class<?> impl) {
+            super(clazz, impl);
+        }
+
+        @Override
+        public Property getProperty(String name) {
+            return name.equals("additionalProperties") ? ADDL_PROPS_PROPERTY : super.getProperty(name);
+        }
+    }
+
+    /**
+     * Property description for an extension subnode.
+     */
+    static class ExtensionProperty extends Property {
+
+        ExtensionProperty(String name) {
+            super(name, Object.class);
+        }
+
+        @Override
+        public Class<?>[] getActualTypeArguments() {
+            return new Class[0];
```

Comment:
Just a quick check to make sure you have to do this, and/or it's the right thing to do.  It struck me as surprising (as a layman reviewer) that the `Property` superclass wouldn't default to this value.

Reply:
In fact, this is correct. We know that the `ExtensionProperty` class will be used only for OpenAPI extensions, and we know those properties will have no type arguments, so this method can safely always return an empty array. I will insert a comment to this effect and will also change the method so it does not create a new array upon each invocation.